DRIVERS-1571 Retry on different mongos when possible (#1450)
mongodb/specifications: source/retryable-reads/tests/README.rst

9. Disable the failpoint.

Retrying Reads in a Sharded Cluster
===================================

These tests will be used to ensure drivers properly retry reads on a different
mongos.

Retryable Reads Are Retried on a Different Mongos if One Available
------------------------------------------------------------------

This test MUST be executed against a sharded cluster that has at least two
mongos instances.

1. Ensure that a test is run against a sharded cluster that has at least two

>    mongos-es. If there are more than two mongos-es in the cluster, pick two to

Reviewer: ```suggestion
   mongoses. If there are more than two mongoses in the cluster, pick two to
```

   test against.

2. Create a client per mongos using the direct connection, and configure fail

>    points on each of the picked mongos-es, so that each mongos raises

Reviewer: ```suggestion
   points on each of the picked mongoses, so that each mongos raises
```

>    a retryable error once.

Reviewer: I suggest noting not to set a failpoint error that is transient, according to the [specifications](https://github.com/mongodb/specifications/blob/master/source/retryable-reads/retryable-reads.rst#why-are-read-operations-only-retried-once-by-default):

>  In the case of a transient error, the driver will mark the server as "unknown"

This could result in false positives. For example, suppose we have two servers: s1, s2.  If s1 fails on the initial attempt, it's type will be set to unknown and is therefore no longer a suitable candidate for server selection (see [here](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#topology-type-unknown)). Thus s2 MUST be selected on the retry and would "fail". In this case, this prose test would pass with or without implementing the mongos retry logic.

3. Create a client with ``retryReads=true`` that connects to the cluster,

>    providing the two selected mongos-es as seeds.

Reviewer: ```suggestion
   providing the two selected mongoses as seeds.
```

4. Enable command monitoring, and execute a read command that is

>    supposed to fail on both mongos-es.

Reviewer: ```suggestion
   supposed to fail on both mongoses.
```

5. Asserts that there were failed command events from each mongos.

6. Disable the fail points.

Retryable Reads Are Retried on the Same Mongos if No Other Available
--------------------------------------------------------------------

1. Ensure that a test is run against a sharded cluster. If there are multiple

>    mongos-es in the cluster, pick one to test against.

Reviewer: ```suggestion
   mongoses in the cluster, pick one to test against.
```

2. Create a client that connects to the mongos using the direct connection,

>    and configure a fail point so that the mongos raises a retryable error once.

Reviewer: I suggest noting not to set a failpoint error that is transient, according to the [specifications](https://github.com/mongodb/specifications/blob/master/source/retryable-reads/retryable-reads.rst#why-are-read-operations-only-retried-once-by-default). See other comment for more details.
